### PR TITLE
fix inferenceservice url with hyphenated knative domainTemplate

### LINKED ISSUE #715: kfserving creates incorrect URL with new Knative serving domainTemplate
/kind bug

**What steps did you take and what happened:**
Knative serving 0.12.1 now supports domainTemplate: https://knative.dev/docs/serving/samples/knative-routing-go/
Kfserving assumes the domainTemplate is '{{.Name}}.{{.Namespace}}.{{.Domain}}' 

Once we update the domainTemplate to '{{.Name}}-{{.Namespace}}.{{.Domain}}', kfseving no longer put namespace in the URL.  

For example, with domainTemplate '{{.Name}}.{{.Namespace}}.{{.Domain}}'  the kfserving URL is `http://sklearn-iris.user1.example.com/v1/models/sklearn-iris` and with domainTemplate '{{.Name}}-{{.Namespace}}.{{.Domain}}', kfserving became ` http://sklearn-iris.example.com/v1/models/sklearn-iris `

**What did you expect to happen:**
kfserving URL should be http://sklearn-iris-user1.example.com/v1/models/sklearn-iris

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Istio Version: 1.1.7
- Knative Version: 0.12.1
- KFServing Version: sha d03512227a90e75aa3a2dc2c449ade942e52321f
- Kubeflow version: N/A
- Minikube version: N/A
- Kubernetes version: (use `kubectl version`): v1.15.5
- OS (e.g. from `/etc/os-release`): ubuntu

### PR DESCRIPTION
**Pull request by yuzliu**
**What this PR does / why we need it**:
Knative serving 0.12.1 now supports domainTemplate: https://knative.dev/docs/serving/samples/knative-routing-go/
Kfserving assumes the domainTemplate is '{{.Name}}.{{.Namespace}}.{{.Domain}}'

Once we update the domainTemplate to '{{.Name}}-{{.Namespace}}.{{.Domain}}', kfseving no longer put namespace in the URL.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #715

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.
no image changes.
**Release note**:

```release-note
Make kfserving compatible with knative domainTemplate
```

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kfserving/717)

### CODE REVIEW

**Comment by ellis-bigelow** [comment]
I'm not totally sold on the idea that we prefer

<isvc_name>-<namespace>.<domain> over <isvc_name>.<namespace>.<domain>

What are the tradeoffs here?

**Comment by yuzisun** [comment]
> I'm not totally sold on the idea that we prefer
> 
> <isvc_name>-. over <isvc_name>..
> 
> What are the tradeoffs here?

We are using a wildcard cert for all our inference services and it does not support multiple levels with default format `isvc_name.namespace-.*`, so we need to change the domain template to make it `isvc_name-namespace-.*`.

File: docs/DEVELOPER_GUIDE.md [hunk 1]
```diff
@@ -39,7 +39,7 @@ Before submitting a PR, see also [CONTRIBUTING.md](../CONTRIBUTING.md).
 
 You must install these tools:
 
-1. [`go`](https://golang.org/doc/install): KFServing controller is written in Go.
+1. [`go`](https://golang.org/doc/install): KFServing controller is written in Go version 1.13.0.
```
Comment (yuzisun) [inline docs/DEVELOPER_GUIDE.md]:
```suggestion
1. [`go`](https://golang.org/doc/install): KFServing controller is written in Go and requires Go 1.13.0+.
```

**Comment by ellis-bigelow** [comment]
> > I'm not totally sold on the idea that we prefer
> > <isvc_name>-. over <isvc_name>..
> > What are the tradeoffs here?
> 
> We are using a wildcard cert for all our inference services and it does not support multiple levels with default format `isvc_name.namespace.domain`, so we need to change the domain template to make it `isvc_name-namespace.domain`.

Is this fundamental to wildcard certs, or specifically the ones you're using? Is it possible to have this format be in the configmap so that you can override it? There are other implications to changing this. Imagine a user wanted to manage a single wildcard cert per namespace. This approach would prevent them from doing so.

File: pkg/constants/constants_test.go [hunk 2]
```diff
@@ -0,0 +1,21 @@
+package constants
+
+import (
+	"github.com/google/go-cmp/cmp"
+	"testing"
+)
+
+func TestVirtualServiceHostname(t *testing.T) {
+
+	expected := "kftest.user1.example.com"
```
[the remark this answers is not in the document]
Comment (yuzliu) [reply pkg/constants/constants_test.go]: Good idea! I updated.

**Comment by yuzliu** [comment]
> > > I'm not totally sold on the idea that we prefer
> > > <isvc_name>-. over <isvc_name>..
> > > What are the tradeoffs here?
> > 
> > 
> > We are using a wildcard cert for all our inference services and it does not support multiple levels with default format `isvc_name.namespace.domain`, so we need to change the domain template to make it `isvc_name-namespace.domain`.
> 
> Is this fundamental to wildcard certs, or specifically the ones you're using? Is it possible to have this format be in the configmap so that you can override it? There are other implications to changing this. Imagine a user wanted to manage a single wildcard cert per namespace. This approach would prevent them from doing so.

@ellis-bigelow it is fundamental to wildcard certs: https://en.wikipedia.org/wiki/Wildcard_certificate 

> The wildcard applies only to one level of the domain name.
> label.label.label.TLD
> *.domain.com is OK. It will match www.domain.com but not domain.com and not 
> zzz.www.domain.com

**Comment by yuzisun** [comment]
> > > I'm not totally sold on the idea that we prefer
> > > <isvc_name>-. over <isvc_name>..
> > > What are the tradeoffs here?
> > 
> > 
> > We are using a wildcard cert for all our inference services and it does not support multiple levels with default format `isvc_name.namespace.domain`, so we need to change the domain template to make it `isvc_name-namespace.domain`.
> 
> Is this fundamental to wildcard certs, or specifically the ones you're using? Is it possible to have this format be in the configmap so that you can override it? There are other implications to changing this. Imagine a user wanted to manage a single wildcard cert per namespace. This approach would prevent them from doing so.

@ellis-bigelow this domain template is in the knative configmap, check the comments here https://github.com/knative/serving/blob/master/config/core/configmaps/network.yaml#L74. We changed the knative domain template configmap and then found this bug because when kfserving creating the top level virtual service it assumes the default template using DOT. So this fix is really remove this assumption to work with any domain template in the configmap.

**Comment by ellis-bigelow** [comment]
> > > > I'm not totally sold on the idea that we prefer
> > > > <isvc_name>-. over <isvc_name>..
> > > > What are the tradeoffs here?
> > > 
> > > 
> > > We are using a wildcard cert for all our inference services and it does not support multiple levels with default format `isvc_name.namespace.domain`, so we need to change the domain template to make it `isvc_name-namespace.domain`.
> > 
> > 
> > Is this fundamental to wildcard certs, or specifically the ones you're using? Is it possible to have this format be in the configmap so that you can override it? There are other implications to changing this. Imagine a user wanted to manage a single wildcard cert per namespace. This approach would prevent them from doing so.
> 
> @ellis-bigelow this domain template is in the knative configmap, check the comments here https://github.com/knative/serving/blob/master/config/core/configmaps/network.yaml#L74. We changed the knative domain template configmap and then found this bug because when kfserving creating the top level virtual service it assumes the default template using DOT. So this fix is really remove this assumption to work with any domain template in the configmap.

Awesome you're all over it! Thanks for clarifying!

File: pkg/constants/constants_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ellis-bigelow) [reply pkg/constants/constants_test.go]: I think this is still missing?

**Comment by ellis-bigelow** [comment]
/lgtm
Just waiting on the table tests for virtualservice_test.go

File: pkg/constants/constants_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (yuzliu) [reply pkg/constants/constants_test.go]: Pushed!

File: Makefile [hunk 3]
```diff
@@ -53,7 +53,13 @@ deploy-dev: manifests
 
 deploy-local: manifests
 	./hack/image_patch_dev.sh local
+	# Remove the certmanager certificate if KFSERVING_ENABLE_SELF_SIGNED_CA is not false
+	cd config/default && if [ ${KFSERVING_ENABLE_SELF_SIGNED_CA} != false ]; then \
+	kustomize edit remove resource certmanager/certificate.yaml; \
+	else kustomize edit add resource certmanager/certificate.yaml; fi;
+
 	kustomize build config/overlays/local | kubectl apply -f -
+	if [ ${KFSERVING_ENABLE_SELF_SIGNED_CA} != false ]; then ./hack/self-signed-ca.sh; fi;
```
Comment (yuzisun) [inline Makefile]: discussed with @yuzliu we will document the local develop in private cloud env.

**Comment by yuzisun** [comment]
thanks @yuzliu!

/lgtm
/approve

**Pull request merged**